File: cressent/modules/gene_map.py
#!/usr/bin/env python3

import argparse
import subprocess
import os
import sys

# Set up argument parser
parser = argparse.ArgumentParser(description='Generate gene arrow plots from motif data using R')

# Required arguments
parser.add_argument('-i', '--input', required=True, help='Input file path for the motif table CSV')

# Optional arguments
parser.add_argument('-o', '--output', default='./', help='Output directory where the plot will be saved')
parser.add_argument('--filename', default='gene_motif.pdf', help='Name of the output file (default: gene_motif.pdf)')
parser.add_argument('--height', type=float, default=10, help='Height of the output plot in inches (default: 10)')
parser.add_argument('--width', type=float, default=10, help='Width of the output plot in inches (default: 10)')
parser.add_argument('-t', '--title', help='Title for the plot (optional)')

def main():
    # Parse arguments
    args = parser.parse_args()
    
    # Ensure filename has a proper extension
    if not os.path.splitext(args.filename)[1]:
        args.filename = args.filename + '.pdf'
        print(f"No file extension detected in filename, using: {args.filename}")
    
    # Create the output directory if it doesn't exist
    if not os.path.exists(args.output):
        print(f"Creating output directory: {args.output}")
        os.makedirs(args.output, exist_ok=True)
    
    # Combine output directory and filename
    output_file_path = os.path.join(args.output, args.filename)
    print(f"Output will be saved to: {output_file_path}")
    
    # Check if input file exists
    if not os.path.exists(args.input):
        print(f"Error: Input file does not exist: {args.input}")
        sys.exit(1)
    
    # Get the path to the R script - assuming it's in the same directory as this script
    r_script_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "gene_map.R")
    
    if not os.path.exists(r_script_path):
        print(f"Error: R script not found at {r_script_path}")
        sys.exit(1)
    
    # Prepare command
    cmd = ["Rscript", r_script_path, args.input, output_file_path, str(args.height), str(args.width)]
    
    # Add title if provided
    if args.title:
        cmd.append(args.title)
    
    print(f"Running command: {' '.join(cmd)}")
    
    try:
        # Run the R script
        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        
        # Stream output in real-time
        for line in process.stdout:
            print(line.strip())
        
        # Wait for the process to complete
        process.wait()
        
        # Check if process was successful
        if process.returncode != 0:
            print("Error: R script execution failed")
            stderr_output = process.stderr.read()
            print(stderr_output)
            sys.exit(process.returncode)
        
        print(f"Successfully created plot: {output_file_path}")
        
    except Exception as e:
        print(f"Error executing R script: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: cressent/modules/gene_map.R
#!/usr/bin/env Rscript

# Load required libraries
suppressPackageStartupMessages({
library(ggplot2)
library(gggenes)
})


# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Function to display usage information
usage <- function() {
  cat("Usage: Rscript gene_map.R [input_file] [output_file] [height] [width] [plot_title]\n")
  cat("\n")
  cat("Arguments:\n")
  cat("  input_file    Path to the motif table CSV (tab-delimited)\n")
  cat("  output_file   Path for the output PDF file (default: gene_motif.pdf)\n")
  cat("  height        Height of the output plot in inches (default: 10)\n")
  cat("  width         Width of the output plot in inches (default: 10)\n")
  cat("  plot_title    Title for the plot (optional)\n")
  cat("\n")
  cat("Example: Rscript gene_map.R motif_table.csv gene_motif.pdf 8 12 \"Motif Distribution\"\n")
  quit(status = 1)
}

# Check if help is requested
if (length(args) == 0 || "--help" %in% args || "-h" %in% args) {
  usage()
}

# Set default values
input_file <- NULL
output_file <- "gene_map_motif"
height <- 10
width <- 10
plot_title <- NULL

# Parse arguments
if (length(args) >= 1) input_file <- args[1]
if (length(args) >= 2) output_file <- args[2]
if (length(args) >= 3) height <- as.numeric(args[3])
if (length(args) >= 4) width <- as.numeric(args[4])
if (length(args) >= 5) plot_title <- args[5]

# Check if input file is provided
if (is.null(input_file)) {
  cat("Error: Input file must be specified.\n\n")
  usage()
}


# Check if the output directory exists, create it if it doesn't
output_dir <- dirname(output_file)
if (!dir.exists(output_dir) && output_dir != ".") {
  cat(paste("Creating output directory:", output_dir, "\n"))
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  if (!dir.exists(output_dir)) {
    cat(paste("Error: Failed to create output directory:", output_dir, "\n"))
    quit(status = 1)
  }
}


# Read and process the data
cat(paste("Reading data from:", input_file, "\n"))
df_genes <- try(read.delim(input_file, sep = "\t"))
if (inherits(df_genes, "try-error")) {
  cat("Error: Failed to read input file. Check if it's a valid tab-delimited file.\n")
  quit(status = 1)
}

# Count rows for legend
n_rows <- length(unique(df_genes$matched))
cat(paste("Found", n_rows, "unique motifs\n"))

# Create the plot
cat("Generating plot...\n")
plot_gene <- ggplot(df_genes, aes(xmin = start, xmax = end, y = seqID, fill = matched)) +
                geom_gene_arrow() +
                facet_wrap(~ seqID, scales = "free", ncol = 1) +
                scale_fill_brewer(palette = "Set3") +
                guides(fill=guide_legend(nrow=n_rows, byrow=TRUE)) +
                labs(y = "Sequence name", fill = "Motifs") +
                theme_genes() +
                theme(legend.position = "bottom")

# Add title if provided
if (!is.null(plot_title)) {
  cat(paste("Adding plot title:", plot_title, "\n"))
  plot_gene <- plot_gene + ggtitle(plot_title) +
    theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))
}

# Save the plot
cat(paste("Saving plot to:", output_file, "\n"))
ggsave(plot_gene, filename = file.path(output_file), height = height, width = width)
cat("Done!\n")

# Rscript /fs/project/PAS1117/ricardo/ssDNA_tool/ssDNA_annotator/modules/gene_map.R \
#                     /fs/project/PAS1117/ricardo/ssDNA_tool/test_data/motif_disc/motif_table.csv \
#                     /fs/project/PAS1117/ricardo/ssDNA_tool/test_data/motif_disc/gene_map_motif.pdf \
#                     10 10 \
#                     "Motif Distribution in Genes"